### PR TITLE
functionalizatinon of cas9_mutant_stats

### PR DESCRIPTION
**Pull request by alexanian**
Based on Chris's work.

Also, just for the public interest, check out some of the new graphs, showing residues that repeat the same amino acid transition across both NGA and NGC klienstiver mutants. Note that residues **1218** and **1219** again are the only frequently-mutated ones that differ between NGA/NGC. Also, we should look for patterns in the ones that only have the same AA substitution once, because maybe they have e.g. equivalent substitutions repeated 3 times.

**for NGA**
aa_mutations

**for NGC**
aa_mutations_ngc

### CODE REVIEW

File: models/tridimensional/data_kleinstiver/aa_info.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+Python variable aa_info describing amino acid properties. Each property contains a dict with keys as the single-letter
+amino acid abbreviations.
+    hydrophobic: string, either 'hydrophobic', 'very hydrophobic', 'neutral', or 'hydrophilic'
+    name: string, full lowercase names for amino acids
+    ph: string indicating ph, either 'neutral', 'acidic' or 'basic'
+    polar: string, either 'polar', 'non-polar' or 'charged'
+    rchain: string, either 'aliphatic', 'aromatic' or 'other'
+    size: string, either 'small', 'medium', 'large', 'large_long chain)' or 'large_benzene)'
+"""
+aa_info = {'hydrophobic': {'A': 'hydrophobic', 'C': 'hydrophobic', 'D': 'hydrophillic', 'E': 'hydrophillic',
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/aa_info.py]: maybe `hydrophobicity`

File: models/tridimensional/data_kleinstiver/aa_info.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+                           'F': 'very hydrophobic', 'G': 'neutral', 'H': 'neutral', 'I': 'very hydrophobic',
+                           'K': 'hydrophillic', 'L': 'very hydrophobic', 'M': 'very hydrophobic', 'N': 'hydrophillic',
+                           'P': 'hydrophillic', 'Q': 'neutral', 'R': 'hydrophillic', 'S': 'neutral', 'T': 'neutral',
+                           'V': 'very hydrophobic', 'W': 'very hydrophobic', 'Y': 'hydrophobic'},
+           'name': {'A': 'alanine', 'C': 'cysteine', 'D': 'aspartic acid', 'E': 'glutamic acid', 'F': 'phenylalanine',
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/aa_info.py:16]: maybe `fullname` or `longname`

File: models/tridimensional/data_kleinstiver/aa_info.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+                    'G': 'glycine', 'H': 'histidine', 'I': 'isoleucine', 'K': 'lysine', 'L': 'leucine',
+                    'M': 'methionine', 'N': 'asparagine', 'P': 'proline', 'Q': 'glutamine', 'R': 'arginine',
+                    'S': 'serine', 'T': 'threonine', 'V': 'valine', 'W': 'tryptophan', 'Y': 'tyrosine'},
+           'ph': {'A': 'neutral', 'C': 'neutral', 'D': 'acidic', 'E': 'acidic', 'F': 'neutral', 'G': 'neutral',
+                  'H': 'basic', 'I': 'neutral', 'K': 'basic', 'L': 'neutral', 'M': 'neutral', 'N': 'neutral',
+                  'P': 'neutral', 'Q': 'neutral', 'R': 'basic', 'S': 'neutral', 'T': 'neutral', 'V': 'neutral',
+                  'W': 'neutral', 'Y': 'neutral'},
+           'polar': {'A': 'non-polar', 'C': 'polar', 'D': 'charged', 'E': 'charged', 'F': 'non-polar', 'G': 'non-polar',
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/aa_info.py]: maybe `polarity`

File: models/tridimensional/data_kleinstiver/aa_info.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+aa_info = {'hydrophobic': {'A': 'hydrophobic', 'C': 'hydrophobic', 'D': 'hydrophillic', 'E': 'hydrophillic',
+                           'F': 'very hydrophobic', 'G': 'neutral', 'H': 'neutral', 'I': 'very hydrophobic',
+                           'K': 'hydrophillic', 'L': 'very hydrophobic', 'M': 'very hydrophobic', 'N': 'hydrophillic',
+                           'P': 'hydrophillic', 'Q': 'neutral', 'R': 'hydrophillic', 'S': 'neutral', 'T': 'neutral',
+                           'V': 'very hydrophobic', 'W': 'very hydrophobic', 'Y': 'hydrophobic'},
+           'name': {'A': 'alanine', 'C': 'cysteine', 'D': 'aspartic acid', 'E': 'glutamic acid', 'F': 'phenylalanine',
+                    'G': 'glycine', 'H': 'histidine', 'I': 'isoleucine', 'K': 'lysine', 'L': 'leucine',
+                    'M': 'methionine', 'N': 'asparagine', 'P': 'proline', 'Q': 'glutamine', 'R': 'arginine',
+                    'S': 'serine', 'T': 'threonine', 'V': 'valine', 'W': 'tryptophan', 'Y': 'tyrosine'},
+           'ph': {'A': 'neutral', 'C': 'neutral', 'D': 'acidic', 'E': 'acidic', 'F': 'neutral', 'G': 'neutral',
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/aa_info.py:20]: ty for not uppercasing that h lol

**Comment by mattsmart** [comment]
aa_info.py makes cas9_mutants.py much cleaner, nice

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 5]
```diff
@@ -28,6 +27,21 @@ def find_mutation_counts(pam):
...
+    Returns a dict w/ the same keys for indices start-end+1 and values = # of times it was mutated for 'pam' = pam
+    """
+    idx_counts = {key: 0 for key in range(start, end+1)}
+    for mutant in mutants_kleinstiver:
+        if mutant['pam'] == pam:
+            for mutation in mutant['mutations']:
+                if start <= mutation['aa_idx'] <= end:
+                    idx_counts[mutation['aa_idx']] += 1
+
+    return idx_counts
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py:40]: no newline above imo

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 6]
```diff
@@ -109,100 +101,67 @@ def find_num_pcr_needed():
...
-        for mutation in mutant['mutations']:
-            aachng = mutation['aa_group_init'] + ' to ' + mutation['aa_group_mut'] + ' at ' + str(mutation['aa_idx'])
-            if aachng in aachng_dict.keys():
-                aachng_dict[aachng] += 1
-            else:
-                aachng_dict[aachng] = 1
-    return aachng_dict
+def hist_mutation_counts(counts_NGA, counts_NGC):
+    """
+    Plots a single figure containing two histograms showing the number of mutations per NGA- or NGA-binding mutant Cas9
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py:109]: NGC or NGA i think

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 7]
```diff
@@ -109,100 +101,67 @@ def find_num_pcr_needed():
...
-    # Get Dictionary of counts
-    ch = aachng_freq(pam)
+    # Sort dictionary & convert to list using custom keys if needed
+    if not sort_key_list:
+        list_to_plot = sorted(dict_to_plot.items())
+    else:
+        list_to_plot = sorted(dict_to_plot.items(), key=lambda x: sort_key_list.index(x[0]))
+
+    # Remove list items with less than min_count
+    if min_count != 1: list_to_plot = [dd for dd in list_to_plot if dd[1] >= min_count]
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: separate the statements

**Comment by mattsmart** [comment]
minor comments, then LGTM. didn't check code logic/output fully, though

**Comment by jrhawley** [comment]
Yeah the dictionary file is long too, so I didn't check all of it, but the structure of it looks good. And I agree with Matt's comments

**Comment by alexanian** [comment]
Made most of requested changes, merging.

**Pull request merged**